sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -83,7 +83,20 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
         numPartitionsSet.headOption
       }
 
-      val targetNumPartitions = requiredNumPartitions.getOrElse(childrenNumPartitions.max)
+      // Read bucketed tables always obeys numShufflePartitions because maxNumPostShufflePartitions
+      // is usually much larger than numShufflePartitions,
+      // which causes some bucket map join lose efficacy after enabling adaptive execution.
+      val nonShuffleChildrenNumPartitions =
+        childrenIndexes.map(children).filterNot(_.isInstanceOf[ShuffleExchangeExec])
+          .map(_.outputPartitioning.numPartitions)
+      val expectedChildrenNumPartitions = if (nonShuffleChildrenNumPartitions.nonEmpty &&
+        conf.maxNumPostShufflePartitions > conf.numShufflePartitions) {
```

Comment:
@cloud-fan @viirya I add `conf.maxNumPostShufflePartitions > conf.numShufflePartitions` to fix test error:
```
org.apache.spark.sql.execution.ReduceNumShufflePartitionsSuite.determining the number of reducers: plan already partitioned(minNumPostShufflePartitions: 5)
org.apache.spark.sql.execution.ReduceNumShufflePartitionsSuite.determining the number of reducers: plan already partitioned
```

https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/113679/testReport/